Custom-ESLint-checks.md

```diff
@@ -0,0 +1,329 @@
+  * [Beyond hello world](#beyond-hello-world)
+    * [Placeholders](#placeholders)
+    * [Navigating the AST](#navigating-the-ast)
+* [Write tests](#write-tests)
+
+## Introduction
+
+We use [ESLint](https://eslint.org) to lint our JavaScript and TypeScript files. While ESLint comes with a ton of built-in rules that we can use, we sometimes need to write our own. This guide will cover how to write these custom ESLint rules and how to test them.
+
+## Write rules
+
+ESLint parses the code being linted into an [abstract syntax tree (AST)](https://en.wikipedia.org/wiki/Abstract_syntax_tree), and our custom rules operate on this AST. To define our rule, we provide functions to be called on various kinds of nodes in the AST. These functions can then report lint errors.
+
+To understand how ASTs and linters work, we recommend that you use [AST Explorer](https://astexplorer.net) to view the AST for your code. You should select "JavaScript" as the language and "babel-eslint" as the parser. Unless you want to write your linter code in AST Explorer, disable the "Transform" option. Your AST Explorer window should look like this:
+
+![A screenshot of AST Explorer with some JavaScript code on the left and the corresponding AST on the right.](images/customEslintChecks/astExplorer2Pane.png)
+
+To write your lint rule in AST Explorer, select "ESLint v4" under the "Transform" option:
+
+![A screenshot of AST Explorer with JavaScript code in the upper left, the corresponding AST in the upper right, a lint rule in the lower left, and the result of running the linter in the lower right](images/customEslintChecks/astExplorer4Pane.png)
+
+### Hello world
+
+To get a better idea for how lint checks work, write a simple ESLint rule that forbids the string `Hello, world!`. Maybe you don't want people accidentally opening PRs that include code that prints `Hello, world!` from when they were learning JavaScript. First consider what kinds of code should violate the rule. Here are some examples:
+
+```js
+const s = "Hello, world!";
+
+console.log("Hello, world!");
+```
+
+#### A note on undecidability
+
+While ESLint parses the code being linted, it does not execute it. This means that no lint rule will be able to prevent `Hello, world!` from being printed. No matter how complicated our lint rule, a devious programmer will always be able to construct some convoluted code that evades the rule. For example, code like this will pass our linter:
+
+```js
+console.log(`Hello, ${"world"}!`)
+```
+
+It turns out that it is provably impossible for us to write a program that can determine whether a program prints `Hello, world!` with perfect accuracy. To see why, suppose for the sake of argument we had a linter `printsHelloWorld()` that accepted the source code of a program and returned a boolean indicating whether the program prints `Hello, world!`. If such a linter existed, we could write the following program, whose source code we'll call `src`:
```

Comment:
I think this argument still holds. Below, I'll try to explain my reasoning a few different ways in the hopes that one will make sense.

> The issue arises because the linter is linting its own code. But it doesn't need to be doing that;

The problem is that a programmer trying to subvert your linter could always write their code to call your linter. If you want your linter to always be correct, then you have to handle the case where the code being linted calls your linter (or includes a copy of your linter code).

> you can think of the linter as an external system operating on the codebase, and the codebase either prints 'Hello, world!' or it does not.

For this reasoning to work, you have to introduce the assumption that the codebase never calls your linter. Therefore your linter is "imperfect" in that there are some programs that it fails on. (Note that my argument never assumes that `printsHelloWorld()` calls the same instance of linter code you are running. It could just be a copy of your linter code.) Further, you have to assume that the codebase doesn't include any code that substantially reproduces the functionality of your linter (imagine that `printsHelloWorld()` is a copy of your linter code with some small changes that affect the linter's behavior on some code, but not on your code). Now we have to define "substantially reproduces," at which point our claim starts to look like "we can write a linter that is always accurate so long as we only lint code that doesn't break the linter," which is compatible with my claim that no linter is perfect.

Here's how a [kubernetes feature proposal doc](https://github.com/kubernetes/enhancements/blob/master/keps/sig-security/1753-logs-sanitization/README.md#theoretical) explains a version of this problem:

> The major theoretical limitation of static analysis is imposed by results from decidability theory - given an arbitrary program written in a general-purpose programming language (one capable of simulating a Turing machine), it is impossible to examine the program algorithmically and determine if an arbitrarily chosen statement in the program will be executed when the program operates on arbitrarily chosen input data ([1]). Furthermore, there is no algorithmic way to identify those programs for which the analysis is possible; otherwise, the halting problem for Turing machines would be solvable.
> ...
> [1] "Tutorial: Static Analysis and Dynamic Testing of Computer ...." https://ieeexplore.ieee.org/abstract/document/1646907/. Accessed 15 May. 2020.

My point here is that we don't have to rely on our intuition that a devious programmer could always write code complex enough to evade our linter. We can actually prove it because the devious programmer could always take include a copy of our linter code in their code. More intuitively, no matter how "fancy" we make our linter, the devious programmer can always make their code even more fancy to subvert our linter. My proof shows this by having the malicious program include a copy of our linter, which you can think of as the malicious program including all the fanciness of the linter, plus some more. This makes the proof easy, but there are endless other ways for the malicious program to be fancier than our linter.

Does that make sense?

Reply:
I don't disagree with the theoretical argument, but I think it entirely misses the point of the linter in the first place. In practical terms, the aim of the linter is to aid reviewers by reducing the need to manually review the easy-to-verify pieces. The linter's aim has never been to catch every single violation and I think doing so is not sensible anyway as that would suggest an obviation of the need for human reviewers.

Therefore I don't think that this argument belongs in the main flow of the doc (though it could be a footnote), and I think it might also be worth making it clear at the outset what the aim of the linter is, so that the question about "can we make a perfect linter" doesn't need to be asked because it's clearly out of scope anyway. That said, I think there is some importance (and that it is feasible) to try and make our linter foolproof for specific, well-defined checks (e.g. things like line length being at most 80 or indentation rules being applied correctly).

Reply:
Ahh okay. That makes sense. I've updated the text to re-frame this discussion as one about what linters are actually capable of